delivery_carrier_return_barcode_pattern/models/delivery_carrier.py

```diff
@@ -0,0 +1,51 @@
+# Copyright 2023 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl)
+
+import re
+
+from odoo import _, api, fields, models
+from odoo.exceptions import UserError
+
+HELP_RETURN_BARCODE_PATTERN = """
+    This pattern can be used to parse a return barcode from the carrier.
+    This has to be a valid regular expression.
+    See https://docs.python.org/3/library/re.html
+"""
+
+
+class DeliveryCarrier(models.Model):
+    _inherit = "delivery.carrier"
+
+    return_barcode_pattern = fields.Char(help=HELP_RETURN_BARCODE_PATTERN)
+
+    @api.constrains("return_barcode_pattern")
+    def _constrains_return_barcode_pattern(self):
+        for record in self:
+            if record.return_barcode_pattern:
+                try:
+                    re.compile(record.return_barcode_pattern)
+                except (re.error):
+                    raise UserError(
+                        _("Return Barcode Pattern must be a valid regular expression")
+                    )
+
+    @api.model
```

Comment:
Shouldn't this method use ormcache ?

Reply:
Hmm.
This depends on the result of `self.search([]).mapped("return_barcode_pattern")`.
Is there a way to do depend on that?
I mean, what happens if a `delivery.carrier.return_barcode_pattern` is edited/added/removed between two calls of this method?
And also, this is quite unlikely that the same barcode would be used multiple times.

Reply:
So, in write(), if return_barcode_pattern is in vals, clear cache.

Reply:
What about this?

> And also, this is quite unlikely that the same barcode would be used multiple times.

As I understand, result would be fetched from the cache only when we already called this method for a given barcode.
When receiving a package for a given carrier, how many times do you need to scan its barcode?
I'd say once.
I'm wondering then if using ormcache here is relevant.
What do you say?

Reply:
Yes indeed.